File: java-kotlin-project/app/src/main/java/example/java/basic/NullSafety.java
package example.java.basic;

import example.java.dataclass.Person;
import example.java.dataclass.User;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class NullSafety {

    private final List<User> userRepository;

    public NullSafety() {
        // userRepository 초기화
        userRepository = new ArrayList<>();
        userRepository.add(new User("john", "john@gmail.com", 25, true));
        userRepository.add(new User("jane", "jane@example.com", 30, false));
        userRepository.add(new User("bob", "bob@test.com", 28, true));
    }

    // 전통적인 null 처리
    public String getUpperCaseName(Person person) {
        if (person != null && person.getName() != null) {
            return person.getName().toUpperCase();
        }

        return "UNKNOWN";
    }

    // Optional 사용 (java 8 이상)
    public Optional<User> findUser(String userId) {
        if (userId == null) {
            return Optional.empty();
        }

        return userRepository.stream()
                .filter(user -> userId.equals(user.getUsername()))
                .findFirst();
    }

    // Optional 체이닝
    public String getUserEmailDomain(String userId) {
        return findUser(userId)
                .map(user -> user.getEmail().split("@")[1])
                .orElse("unknown");
    }
}

File: java-kotlin-project/app/src/main/java/example/kotlin/basic/NullSafety.kt
package example.kotlin.basic

import example.kotlin.dataclass.Person
import example.kotlin.dataclass.User

class NullSafety {

    // Nullable 타입과 safe call
    fun getUpperCaseName(person: Person?): String {
        return person?.name?.uppercase() ?: "UNKNOWN"
    }

    // Elvis 연산자(?:)로 기본값 제공
    fun getUserDisplayName(user: User?): String {
        return user?.username ?: "Anonymous"
    }

    // Safe call 체이닝
    fun getUserEmailDomain(user: User?): String {
        // return user?.email?.substringAfter("@") ?: "unknown" // email이 빈 문자열일 때 null이 아니라서 의도한 대로 동작하지 않았음
        return user?.email?.takeIf { it.contains("@") }?.substringAfter("@") ?: "unknown"
    }

    // let을 이용한 null 처리
    fun processUser(user: User?) {
        user?.let { u ->
            println("Processing user: ${u.username}")
            println("Email: ${u.email}")
        }
    }

    // 강제 언랩핑(!!) - 주의
    fun riskyFunction(text: String?) {
        val length = text!!.length // null이면 NullPointerException
    }

    // 스마트 캐스팅
    fun smartCastExample(text: String?) {
        if (text != null) { // 이 블록 안에서는 text가 자동으로 String으로 캐스팅
            println(text.length)    // ?. 없이 사용 가능
        }
    }
}